db/version_set.h

```diff
@@ -134,7 +135,8 @@ class VersionStorageInfo {
                      bool _force_consistency_checks,
                      EpochNumberRequirement epoch_number_requirement,
                      SystemClock* clock,
-                     uint32_t bottommost_file_compaction_delay);
+                     uint32_t bottommost_file_compaction_delay,
+                     const OffpeakTimeInfo& offpeak_time_info);
```

Comment:
For a function or ctor that is always saving a copy of an object, AFAIK, of the three options passing {by value, by const ref and by rval ref (`&&`)}, the worst choice is passing by const ref. AFAIK, unless the function is inlined, you're forcing a copy to be made even if the caller is done with its version of the object (e.g. temporary) because the copy is made in the callee. With the other two options, the caller should be able to avoid the copy in many cases.

(An anti-pattern example of this that I have pointed out to others before is `const std::shared_ptr<X>&` parameter on a function or ctor that almost always saves a copy of the shared_ptr. Simple `std::shared_ptr<X>` is better if the callee is saving a copy.)

See https://en.cppreference.com/w/cpp/language/copy_elision

Reply:
If my understanding is correct, when we pass by value, compiler is smart enough to avoid copying the object twice? (maybe something similar to NRVO when it's created to a place needed by the caller?)

Or I should just use `std::move` if I don't need it from the caller side (but this isn't my option since I cannot lose it from `VersionSet`)